### ISSUE TITLE
Remove the `PREDICT` macros.

### ISSUE THREAD

**Issue by markshannon** [python/cpython#104610]
The `PREDICT()` macros are supposed to speed dispatch on platforms without computed gotos (i.e. Windows).
However the "predictions" are nothing like the actual [behavior observed](https://github.com/faster-cpython/ideas/blob/main/stats/pystats-2023-05-14-python-7d2deaf.md#pair-counts), especially with PEP 659.

We might want to insert something like the `PREDICT` macros automatically, but the manually inserted ones are useless, or worse, and should be removed.

### Linked PRs
* gh-104651
* gh-105459

**Comment by brandtbucher** [comment]
See previous discussion at faster-cpython/ideas#496.

So we should probably look at comparing Windows benchmarking results before doing this (which is easy, since we have the machine now).

**Comment by markshannon** [comment]
Those are the Windows benchmarking results.

**Comment by rhettinger** [comment]
> The PREDICT() macros are supposed to speed dispatch on platforms with computed gotos

The macros sped dispatch on platforms and builds where the big case-statement made unpredictable branches.  At the time the macros were introduced, the speed-up was significant because the PREDICT macros made branches that were highly predictable while the other indirect branches where unpredictable.

Subsequently, some processors implemented more advanced branch prediction logic that included indirect branches.  To exploit that capability, the computed-goto option was added.  Eventually that option became the default as other processors got smarter.  We left the macros in because people were still occasionally doing 32-bit builds and sometimes targeting non-intel processors.

In short, the benefits of the PREDICT macro are processor and build specific.  We've long known that with computed gotos enabled on a good processor that benchmarks would show no benefit (i.e. you haven't found anything new, that is why the default build was changed).  That macros are there help in the remaining situations where the case-statement is still making unpredictable branches, not because of a build option, but because the processor itself has weak branch prediction.

**Comment by markshannon** [comment]
There is no benefit to the PREDICT macros at the moment. The predictions are wrong. They do not help; they hinder.

Hypothetically, they could be useful if the predictions were accurate, but we would need to generate them from profiling data for that to happen.

**Comment by markshannon** [comment]
I realize that I originally wrote "with computed gotos" rather than "without computed gotos".
I can see how that could be confusing. Sorry about that.

On platforms *without* computed gotos, prediction macros have the *potential* to help dispatch, but in practice, they don't.
